### PR TITLE
feat: search ui debounce #2472

### PR DESCRIPTION
**Pull request by frolyo**

### CODE REVIEW

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    protected override Task DisposeAsyncCore()
+    {
+        _queue.Writer.TryComplete();
+        return base.DisposeAsyncCore();
+    }
+
+    public void Enqueue(T item)
+    {
+        if (!_queue.Writer.TryWrite(new ActionContext(item, Clock.Now)))
+            throw StandardError.Internal("Failed to enqueue debounced task");
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:39]: Dot is missing in the end of exception message.

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                var delay = Interval - (clockNow - context.QueuedAt);
+                if (delay > TimeSpan.Zero)
+                    await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+                if (_queue.Reader.Count == 0) {
+                    await executing.DisposeSilentlyAsync().ConfigureAwait(false);
+                    executing = context;
+                    executing.Start(ActionFactory);
+                }
+                else {
+                    Console.WriteLine("Not empty");
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:57]: Better use logger here.

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+public class CancellableDebouncer<T> : WorkerBase
+{
+    private readonly Channel<ActionContext> _queue = Channel.CreateBounded<ActionContext>(new BoundedChannelOptions(100) {
+        FullMode = BoundedChannelFullMode.DropOldest,
+    });
+
+    private MomentClock Clock { get; }
+    private ILogger Log { get; }
+    private TimeSpan Interval { get; }
+    private Func<T, CancellationToken, Task> ActionFactory { get; }
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:12]: may better name prop: TaskFactory ?

**Review by alexis-kochetov — APPROVED** [review]

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                if (_queue.Reader.Count == 0) {
+                    await executing.DisposeSilentlyAsync().ConfigureAwait(false);
+                    executing = context;
+                    executing.Start(ActionFactory);
+                }
+                else {
+                    Console.WriteLine("Not empty");
+                }
+            }
+            catch (Exception e) {
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:60]: I think we should not log OperationCancelledException

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (_task != null)
+                await _task.WaitAsync(TimeSpan.FromSeconds(1), CancellationToken.None)
+                    .SuppressExceptions()
+                    .ConfigureAwait(false);
+        }
+    }
+}
+
+public class CancellableDebouncer(MomentClock clock, ILogger log, TimeSpan interval, Func<CancellationToken, Task> actionFactory)
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:114]: I would prefer to have Debouncer class with static factory methods to cover different scenarious of Debouncing and Throttling than another one type.

File: src/dotnet/Core/IO/CancellableDebouncer.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
+namespace ActualChat.IO;
+
+public class CancellableDebouncer<T> : WorkerBase
```
Comment (crui3er) [inline src/dotnet/Core/IO/CancellableDebouncer.cs:3]: It's not obvious when to use CancellableDebouncer<T> and when Debouncer.

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 1 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:39]: fixed

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 2 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:57]: removed this dubug code

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 4 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:60]: done

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 3 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:12]: done

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 5 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:114]: done

File: src/dotnet/Core/IO/CancellableDebouncer.cs (hunk 6 above)
Comment (frolyo) [reply src/dotnet/Core/IO/CancellableDebouncer.cs:3]: added Debouncer.New...

**Pull request merged**